File: java/p215.java
/* 
 * Solution to Project Euler problem 215
 * Copyright (c) Project Nayuki. All rights reserved.
 * 
 * https://www.nayuki.io/page/project-euler-solutions
 * https://github.com/nayuki/Project-Euler-solutions
 */

import java.math.BigInteger;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Stack;


public final class p215 implements EulerSolution {
	
	public static void main(String[] args) {
		System.out.println(new p215().run());
	}
	
	
	private static final int WIDTH = 32;
	private static final int HEIGHT = 10;
	
	
	public String run() {
		List<int[]> crackPositions = new ArrayList<>();
		getCrackPositions(new Stack<Integer>(), 0, crackPositions);
		
		BigInteger[] ways = new BigInteger[crackPositions.size()];
		Arrays.fill(ways, BigInteger.ONE);
		
		for (int i = 1; i < HEIGHT; i++) {
			BigInteger[] newWays = new BigInteger[ways.length];
			for (int j = 0; j < newWays.length; j++) {
				BigInteger sum = BigInteger.ZERO;
				for (int k = 0; k < ways.length; k++) {
					if (areDisjointSorted(crackPositions.get(j), crackPositions.get(k)))
						sum = sum.add(ways[k]);
				}
				newWays[j] = sum;
			}
			ways = newWays;
		}
		
		BigInteger sum = BigInteger.ZERO;
		for (BigInteger x : ways)
			sum = sum.add(x);
		return sum.toString();
	}


	private static void getCrackPositions(Stack<Integer> cracks, int position, List<int[]> result) {
		if (position < 0)
			throw new IllegalArgumentException();
		else if (position < WIDTH) {
			for (int i = 2; i <= 3; i++) {
				cracks.push(position + i);
				getCrackPositions(cracks, position + i, result);
				cracks.pop();
			}
		} else if (position == WIDTH) {
			int[] temp = new int[cracks.size() - 1];
			for (int i = 0; i < temp.length; i++)
				temp[i] = cracks.get(i);
			result.add(temp);
		} else  // position > WIDTH
			return;
	}
	
	
	private static boolean areDisjointSorted(int[] a, int[] b) {
		for (int i = 0, j = 0; i < a.length && j < b.length; ) {
			if (a[i] == b[j])
				return false;
			else if (a[i] < b[j])
				i++;
			else if (a[i] > b[j])
				j++;
			else
				throw new AssertionError();
		}
		return true;
	}
	
}

File: python/p215.py
# 
# Solution to Project Euler problem 215
# Copyright (c) Project Nayuki. All rights reserved.
# 
# https://www.nayuki.io/page/project-euler-solutions
# https://github.com/nayuki/Project-Euler-solutions
# 


def compute():
	WIDTH = 32
	HEIGHT = 10
	crackpositions = []
	
	def get_crack_positions(cracks, position):
		if position < 0:
			raise ValueError()
		elif position < WIDTH:
			for i in (2, 3):
				cracks.append(position + i)
				get_crack_positions(cracks, position + i)
				cracks.pop()
		elif position == WIDTH:
			crackpositions.append(frozenset(cracks[ : -1]))
		else:  # position > WIDTH
			return
	
	get_crack_positions([], 0)
	
	noncrackindices = [
		[i for (i, cp1) in enumerate(crackpositions) if cp0.isdisjoint(cp1)]
		for cp0 in crackpositions]
	
	ways = [1] * len(crackpositions)
	for i in range(1, HEIGHT):
		newways = [sum(ways[k] for k in nci) for nci in noncrackindices]
		ways = newways
	
	ans = sum(ways)
	return str(ans)


if __name__ == "__main__":
	print(compute())
